### PR TITLE
FreeBSD compatibility of 3.8 - fixes #1138

### LINKED ISSUE #1138: 3.8 beta1 fails to compile on FreeBSD
When trying to compile 3.8 beta1 on FreeBSD, it fails with the following message:
```
  CC       modules/date/modules_date_libdate_la-date-parser-parser.lo
  CC       modules/date/modules_date_libdate_la-strptime-tz.lo
modules/date/strptime-tz.c:516:19: error: invalid argument type 'char *(*)(int, int)' to unary expression
                                        *tm_gmtoff = -(timezone);
                                                     ^~~~~~~~~~~
1 error generated.
gmake[4]: *** [Makefile:10516: modules/date/modules_date_libdate_la-strptime-tz.lo] Error 1
gmake[3]: *** [Makefile:13201: all-recursive] Error 1
gmake[2]: *** [Makefile:5440: all] Error 2
gmake[2]: Leaving directory '/root/syslog-ng-devel/work/syslog-ng-3.8.0beta1'
*** Error code 1
```

### PR DESCRIPTION
**Pull request by kvch**
Two problems were detected on FreeBSD
- timezone function call incompatibility
- missing header includes in two files

Now, syslog-ng compiles and make check passes. However, I haven't tested syslog-ng with python modules and java modules yet.

### CODE REVIEW

File: modules/date/strptime-tz.c [hunk 1]
```diff
@@ -513,7 +513,14 @@ again:		switch (c = *fmt++) {
...
 					tm->tm_isdst = i;
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+#include <sys/param.h>
+#if defined(BSD)
+					*tm_gmtoff = -(tm->tm_gmtoff);
+#else
 					*tm_gmtoff = -(timezone);
+#endif
+#endif
 					*tm_zone = tzname[i];
```
Comment (bazsi) [inline modules/date/strptime-tz.c]: This should be extracted into a new function.

long _get_gmtoff_from_struct_tm(const struct tm *tm);

**Comment by bazsi** [comment]
apart from the timezone() stuff, the rest is :+1:

**Comment by kvch** [comment]
I fixed the code based on @bazsi's note. But unfortunately, I haven't been able to make the CI green yet. I might need some help with it.

**Comment by bazsi** [comment]
You seem to have two shadowed identifiers in strptime-tz.c, e.g. you have
identifiers that have the same name as a global declaration.

**Comment by czanik** [comment]
I just tried to compile beta1 with https://patch-diff.githubusercontent.com/raw/balabit/syslog-ng/pull/1141.patch on FreeBSD 10.3 but it failed:
```
  CC       modules/date/modules_date_libdate_la-date-parser.lo
  CC       modules/date/modules_date_libdate_la-date-parser-parser.lo
  CC       modules/date/modules_date_libdate_la-strptime-tz.lo
modules/date/strptime-tz.c:206:18: error: use of undeclared identifier 'tm'
  *tm_gmtoff = -(tm->tm_gmtoff);
                 ^
1 error generated.
gmake[4]: *** [Makefile:10516: modules/date/modules_date_libdate_la-strptime-tz.lo] Error 1
gmake[3]: *** [Makefile:13201: all-recursive] Error 1
gmake[2]: *** [Makefile:5440: all] Error 2
gmake[2]: Leaving directory '/root/syslog-ng-devel/work/syslog-ng-3.8.0beta1'
*** Error code 1
```

**Comment by kvch** [comment]
@czanik /o\ I'll look into it.

**Comment by czanik** [comment]
Thanks, core package compiles now.

**Comment by kvch** [comment]
@czanik Cool. But I cannot state that my fix really solves the problem. Date-parser is still problematic.

File: modules/date/strptime-tz.c [hunk 2]
```diff
@@ -88,6 +88,15 @@ typedef unsigned int uint;
 #define isleap(y) ((((y) % 4) == 0 && ((y) % 100) != 0) || ((y) % 400) == 0)
 #define isleap_sum(a, b)	isleap((a) % 400 + (b) % 400)
 
+#define GET_OFFSET_FROM_UTC *tm_gmtoff = -(timezone)
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+#include <sys/param.h>
+#if defined(BSD)
+#define GET_OFFSET_FROM_UTC -(tm->tm_gmtoff)
```
Comment (lbudai) [inline modules/date/strptime-tz.c]: This won't work (missing lvalue).

**Comment by lbudai** [comment]
@kvch : I have this fix
https://github.com/balabit/syslog-ng/compare/master...lbudai:f/strptime-with-tz-timezone-fix

I'll create a PR to your branch and then your PR will be ready for merge.

**Pull request merged**